[Cocoa][MSE] Fix canvas painting on older builds

youssefsoli opened the pull request:
#### 02bb8a509f08174d0673b37f1ceef03cfd689824
<pre>
[Cocoa][MSE] Fix canvas painting on older builds
<a href="https://bugs.webkit.org/show_bug.cgi?id=243163">https://bugs.webkit.org/show_bug.cgi?id=243163</a>
&lt;rdar://97532406&gt;

Reviewed by Eric Carlson and Wenson Hsieh.

After 252183@main, an alternate path was provided for older builds that
didn&apos;t support AVSBDL `-copyDisplayedPixelBuffer` by calling `willBeAskedToPaintGL`
before a canvas paint is attempted. The problem is that the location in
which `willBeAskedToPaintGL` was called is too late since valid frame metadata
is checked before `willBeAskedToPaintGL` is reached, however, on older builds
this method needs to be called in order to valid frame metadata to be generated.

The fix is to simply enable the decompression session if frame metadata
is being requested and notify WebGL is about to paint for the SW path as well.

* Source/WebCore/html/canvas/WebGLRenderingContextBase.cpp:
(WebCore::WebGLRenderingContextBase::texImageSourceHelper):
* Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateMediaSourceAVFObjC.mm:
(WebCore::MediaPlayerPrivateMediaSourceAVFObjC::shouldEnsureLayer const):
(WebCore::MediaPlayerPrivateMediaSourceAVFObjC::startVideoFrameMetadataGathering):
(WebCore::MediaPlayerPrivateMediaSourceAVFObjC::stopVideoFrameMetadataGathering):
* Source/WebCore/platform/graphics/cocoa/MediaPlayerPrivateWebM.mm:
(WebCore::MediaPlayerPrivateWebM::shouldEnsureLayer const):
(WebCore::MediaPlayerPrivateWebM::startVideoFrameMetadataGathering):
(WebCore::MediaPlayerPrivateWebM::stopVideoFrameMetadataGathering):

* LayoutTests/platform/mac/TestExpectations:

Canonical link: <a href="https://commits.webkit.org/252893@main">https://commits.webkit.org/252893@main</a>
</pre>

youennf commented on Source/WebCore/html/HTMLVideoElement.cpp:
```diff
@@ -675,8 +679,12 @@ void HTMLVideoElement::serviceRequestVideoFrameCallbacks(ReducedResolutionSecond
     HTMLMediaElement::mediaPlayerEngineUpdated();
-    if (!m_videoFrameRequests.isEmpty() && player())
+    if (!m_videoFrameRequests.isEmpty() && player()) {
+#if PLATFORM(COCOA) && !HAVE(AVSAMPLEBUFFERDISPLAYLAYER_COPYDISPLAYEDPIXELBUFFER)
```
Why not calling willBeAskedToPaintGL() in the startVideoFrameMetadataGathering() specialisation?

youssefsoli replied:
Will there be a scenario where `startVideoFrameMetadataGathering()` is called without the need for painting? Otherwise, it makes sense to plop it in there to prevent the extra IPC call.

youennf replied:
Some websites compute frame rates from rvfc for logging/qa metrics purposes.

youssefsoli replied:
In that case, this change will break those websites as @whsieh mentioned since the display layer will be destroyed on rvfcv for older builds.

eric-carlson commented on Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateMediaSourceAVFObjC.mm:
```diff
@@ -1431,6 +1431,10 @@ void getSupportedTypes(HashSet<String, ASCIICaseInsensitiveHash>& types) const f
     m_isGatheringVideoFrameMetadata = true;
+    
+#if PLATFORM(COCOA) && !HAVE(AVSAMPLEBUFFERDISPLAYLAYER_COPYDISPLAYEDPIXELBUFFER)
+    willBeAskedToPaintGL();
```
`willBeAskedToPaintGL` makes this work as a side effect. I would prefer if we had a new method with a name that made sense for what it does, even if it does exactly the same thing.

youssefsoli committed:
[Cocoa][MSE] Fix canvas painting on older builds
https://bugs.webkit.org/show_bug.cgi?id=243163
<rdar://97532406>

Reviewed by Eric Carlson and Wenson Hsieh.

After 252183@main, an alternate path was provided for older builds that
didn't support AVSBDL `-copyDisplayedPixelBuffer` by calling `willBeAskedToPaintGL`
before a canvas paint is attempted. The problem is that the location in
which `willBeAskedToPaintGL` was called is too late since valid frame metadata
is checked before `willBeAskedToPaintGL` is reached, however, on older builds
this method needs to be called in order to valid frame metadata to be generated.

The fix is to simply enable the decompression session if frame metadata
is being requested and notify WebGL is about to paint for the SW path as well.

* Source/WebCore/html/canvas/WebGLRenderingContextBase.cpp:
(WebCore::WebGLRenderingContextBase::texImageSourceHelper):
* Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateMediaSourceAVFObjC.mm:
(WebCore::MediaPlayerPrivateMediaSourceAVFObjC::shouldEnsureLayer const):
(WebCore::MediaPlayerPrivateMediaSourceAVFObjC::startVideoFrameMetadataGathering):
(WebCore::MediaPlayerPrivateMediaSourceAVFObjC::stopVideoFrameMetadataGathering):
* Source/WebCore/platform/graphics/cocoa/MediaPlayerPrivateWebM.mm:
(WebCore::MediaPlayerPrivateWebM::shouldEnsureLayer const):
(WebCore::MediaPlayerPrivateWebM::startVideoFrameMetadataGathering):
(WebCore::MediaPlayerPrivateWebM::stopVideoFrameMetadataGathering):

* LayoutTests/platform/mac/TestExpectations:

Canonical link: https://commits.webkit.org/252893@main
